### ISSUE TITLE
CARDspa

### ISSUE THREAD

**Issue by JINGFU9912** [Bioconductor/Contributions#3470]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/YMa-lab/CARDspa

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by lshep** [comment]
set.seed should not be used within R functions themselves but called outside the function.   If there is some strong reasoning for using within R functions minimally it should be controlled by an argument to the function that the user can control

**Comment by JINGFU9912** [comment]
Hi @lshep,

Thanks for your comments. I have updated the code to allow users the option to control the seed setting. Here's an explanation of why I use set.seed() in my code:

R/CARD.prop.R (line 268), R/CARD.refFree.R (lines 73 and 86), R/CARD.SCMapping.R (line 64):
The use of set.seed() in these functions is intended to ensure consistent parameter initialization across different runs. Initial conditions can significantly affect the results obtained by the user.

R/visualization.R (line 318):
Here, set.seed() is used to generate a consistent set of colors for visualizations.

Best regards,
Jing

**Comment by JINGFU9912** [comment]
Hi! I wanted to kindly ask if there are any updates or feedback on the review for our package?

**Comment by ococrook** [comment]
Hi!

Sorry for the delay September has been very busy. Looking through the package it is applied to spatial transcriptomic data. I'm not an expert on this data type but we do have a core principle of re-using Bioconductor infrastructure. Is there any reason that input type cannot be a `spatialExperiment` object. The code might require a bit of rewriting to get that to work.

Best wishes,
Olly

**Comment by JINGFU9912** [comment]
> Hi!
> 
> Sorry for the delay September has been very busy. Looking through the package it is applied to spatial transcriptomic data. I'm not an expert on this data type but we do have a core principle of re-using Bioconductor infrastructure. Is there any reason that input type cannot be a `spatialExperiment` object. The code might require a bit of rewriting to get that to work.
> 
> Best wishes, Olly

Hi Olly,

Thank you for your comments and for taking the time to review CARDspa. I understand the importance of adhering to Bioconductor's principle of reusing existing infrastructure, and I appreciate your suggestion regarding the use of the SpatialExperiment object.

The current design of CARDspa aims to provide flexibility by allowing users to input single-cell and spatial genomics data directly as matrices. This was done to simplify data preparation, especially for users who may not be familiar with SingleCellExperiment or SpatialExperiment.

If we were to switch to SpatialExperiment, it would require users to convert single-cell data into SingleCellExperiment objects as well. Since spatial genomics data typically comes in the form of separate count matrices and spatial coordinates, enforcing the use of SpatialExperiment could add unnecessary complexity for users, who would need to preprocess their data. Furthermore, CARDspa would need to extract the count matrices and coordinates from within the SpatialExperiment object, adding redundant steps.

Best Wishes,
Jing

**Comment by vjcitn** [comment]
Thanks for your submission.  
```
> getClass("CARD")
Class "CARD" [package "CARDspa"]

Slots:
                                                               
Name:             sc_eset   spatial_countMat   spatial_location
Class:                ANY                ANY         data.frame
                                                               
Name:     Proportion_CARD            project    info_parameters
Class:             matrix          character               list
                                                               
Name:    algorithm_matrix       refined_prop refined_expression
Class:               list             matrix             matrix
```
You've created an S4 class that shares certain features with SpatialExperiment, but lacks a show method, so printing it leads to screenfuls of raw information.  

Our guidelines indicate the importance of supporting interop with *existing* Bioconductor infrastructure.  While you remark that 
```
enforcing the use of SpatialExperiment could add unnecessary complexity for users
```
I would have a couple of responses.  First, we don't enforce use of anything.  Reviews and guidelines may nudge
developers in certain directions.  Second, I don't know what "unnecessary complexity" is, but printing a CARD object
is currently making the user confront some serious complexity.  Third, in our view and in the view of contributing developers, *avoiding* SpatialExperiment and derivates produces more complexity than using it.  This looks like a great package and I hope it gets wide use.  If you don't want to use Bioconductor classes, that's fine, if you contribute it to CRAN, Bioconductor packages can import it and make use of it.  If that's the direction you wish to pursue, please close the issue.  If you wish to receive a review at Bioconductor, please do implement interop with existing classes.

**Comment by JINGFU9912** [comment]
Hi @vjcitn and @ococrook ,

Thank you for your valuable feedback and I already uploaded the latest version.

1. Based on your suggestion, I have added a show method. Now, when users print the CARD object, they will no longer see excessive raw data output, enhancing usability.

2. I sincerely appreciate your explanation regarding the use of SpatialExperiment and Bioconductor classes. I understand and agree with your concerns, and I believe this approach better promotes consistency within the Bioconductor ecosystem. Thus, I have modified the code to effectively interact with SpatialExperiment and SingleCellExperiment objects. Additionally, I have added a vignette to demonstrate this functionality, ensuring clarity and ease of use.

Please let me know if there are any further comments or suggestions. Thanks again!

Best,
Jing

**Comment by ococrook** [comment]
Hi Jing

Thanks for the changes, I've been through the code and have some suggestions for improvement::

- [ ] Would benefit from a readme
- [ ] Please improve unit testing coverage
- [ ]  Check notes in the build - some are very easy to fix, in particular the sapply note, and the use of the colon to generate sequences e.g. use seq functions instead
- [ ]  I wouldn’t use “x” to reference an object please use descriptive input names e.g object, params, SpatialExp, etc
- [ ] Please check inputs are reasonable e.g. having a “stopifnot” statement that check the inputs are valid
- [ ]  Remove code that has been commented out, if necessary to keep please isolate away from the structure of the functioning code 
- [ ]  Your parallelization approach breaks on windows, could you allow users to specify the backend that will work for them or make sure this won’t be platform dependent
- [ ]  Could you move classes to AllClasses.R - it took me a while to locate them.  
- [ ]  I’m not sure whether the “SpatialExperiment” object conversion to CARD is exactly what was intended by interoperability. Yes, your functions will work but would it not be better just to use these object straight-up and use the CARD object behind the scenes? Would appreciate a second opinion from @vjcitn 
- [ ] I generally found the code difficult to follow as there are a number of standard coding practices aren't kept to I would recommend:
A) avoid re-using the same variable to have different meanings in the same chunk of code 
B) avoid complex mutil-logical statements in the same line e.g. one-line - one -purpose would be benifical
C) Employ better use of white-space. Try to connect logical pieces of code together with a prevailing comment e.g. storage, statistical computations, filter etc 
D) You switch styles between using ".", "_" and camelCase. In general use of capital letters is inconsistent, which makes it hard to follow. The most important thing is to be consistent but would suggest having a look the the Bioconductor style guide suggestions

**Comment by lshep** [comment]
We are updating our builders;  I believe this to be an error on our end that we are investigating

**Comment by JINGFU9912** [comment]
Hi @lshep,

Thank you for letting me know. I was thinking how to address this error, so your clarification is much appreciated. Please let me know if there is anything I need to do on my end in the meantime.

-Jing

**Comment by lshep** [comment]
I will kick off a new build for you once we have completed updates

**Comment by JINGFU9912** [comment]
Thank you so much @lshep

**Comment by JINGFU9912** [comment]
Hi @ococrook ,

Thank you for your valuable feedback, I have made all the changes.

- [x] Would benefit from a readme
- [x] Please improve unit testing coverage
- [x] Check notes in the build - some are very easy to fix, in particular the sapply note, and the use of the colon to generate sequences e.g. use seq functions instead
- [x] I wouldn’t use “x” to reference an object please use descriptive input names e.g object, params, SpatialExp, etc
- [x] Please check inputs are reasonable e.g. having a “stopifnot” statement that check the inputs are valid
- [x] Remove code that has been commented out, if necessary to keep please isolate away from the structure of the functioning code
- [x] Your parallelization approach breaks on windows, could you allow users to specify the backend that will work for them or make sure this won’t be platform dependent
- [x] Could you move classes to AllClasses.R - it took me a while to locate them.
- [x] I’m not sure whether the “SpatialExperiment” object conversion to CARD is exactly what was intended by interoperability. Yes, your functions will work but would it not be better just to use these object straight-up and use the CARD object behind the scenes? Would appreciate a second opinion from @vjcitn
- [x] I generally found the code difficult to follow as there are a number of standard coding practices aren't kept to I would recommend:
- [x] A) avoid re-using the same variable to have different meanings in the same chunk of code
- [x] B) avoid complex mutil-logical statements in the same line e.g. one-line - one -purpose would be benifical
- [x] C) Employ better use of white-space. Try to connect logical pieces of code together with a prevailing comment e.g. storage, statistical computations, filter etc
- [x] D) You switch styles between using ".", "_" and camelCase. In general use of capital letters is inconsistent, which makes it hard to follow. The most important thing is to be consistent but would suggest having a look the the Bioconductor style guide suggestions

-----------------------------------
I have made the following changes based on your suggestions:

-Added a README file.
-Added unit tests for all functions using testthat.
-Addressed all resolvable notes from the build report.
-Renamed variables to be more meaningful.
-Implemented input validation with stopifnot checks.
-Removed commented-out code.
-Used BiocParallel to ensure parallelization works on Windows.
-Moved all class definitions to AllClasses.R.
-Updated functions so that the CARD object is used internally, but the final output is a SpatialExperiment object for better user understanding.
-Refactored code to improve readability and adhere to a consistent style, following Bioconductor guidelines.

Thank you again for your suggestions.

-Jing

**Comment by lshep** [comment]
@ococrook  will you be doing a re-review here?

**Comment by ococrook** [comment]
Hi @JINGFU9912 

Thanks for the changes - though I dont seem them reflected in your github repo? Did you forget to push?

Best,
Olly

**Comment by JINGFU9912** [comment]
Thanks for the reminder, I have already pushed it. @ococrook

**Comment by ococrook** [comment]
Great - I was confused at first. This is much improved thank you! Could you include a NEWS.md file to keep track of changes between version and then bump the version number to 0.99.5 in the description and NEWS file.  You can see a NEWS.md file example here: https://github.com/ococrook/bandle/blob/main/NEWS.md

For 0.99.5 you can write bioc ready, and 0.99.4 summarise the changes you made. Youll need to keep the updates as changes are made to the package with version numbers.

**Comment by JINGFU9912** [comment]
Hi @ococrook, thank you for your suggestion! I have added the NEWS.md file and bumped the version to 0.99.5 as requested.

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/JINGFU9912.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("CARDspa")`. The package 'landing page' will be created at

https://bioconductor.org/packages/CARDspa

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
